src/scripts/circleci-job-canceller/long_job_canceller.py

```diff
@@ -60,50 +60,63 @@ def pipeline_created_at_to_datetime(pipeline):
 
 
-def find_old_workflow_ids(repo_slug, window_start, window_end, headers):
-    print(f'Window to paginate through: [{window_start}, {window_end}]')
+def find_old_workflow_ids(
+        repo_slug,
+        window_start,
+        window_end_cancel,
+        window_end_warn,
+        headers):
+    print(
+        f'Window to paginate through: [start:{window_start}, cancel:{window_end_cancel}, warn:{window_end_warn}]')
     for current_pipeline in get_all_items(f"/project/gh/{repo_slug}/pipeline", headers, None):
         # Paginate through only those pipelines which started inside our given window
         created_at = pipeline_created_at_to_datetime(current_pipeline)
-        if window_end < created_at:
+        if window_end_warn < created_at:
             print(f'Pipeline too young: {created_at}')
             continue
         if created_at < window_start:
             print(f'Pipeline too old: {created_at}')
             return None
 
         for current_workflow in get_all_items(f"/pipeline/{current_pipeline['id']}/workflow", headers, None):
-            if current_workflow["status"] == "on_hold":
-                created_at_str = current_workflow["created_at"]
-                created_at = isoparse(created_at_str)
-
-                if ((created_at < (now - job_midlife_warning)) and (created_at > (now - job_life_clock))):
-                    username = get_workflow_started_by(
-                        current_workflow, headers)
-                    if username in robot_committers:
-                        continue
+            job_status = None
+            username = get_workflow_started_by(current_workflow, headers)
+            logging_detail = f'[{created_at}] ({current_workflow["name"]})[{current_workflow["status"]}], started by gh:{username}. See more info at: https://app.circleci.com/pipelines/workflows/{current_workflow["id"]}'
+
+            if not current_workflow.get('stopped_at'):
+                if (created_at < window_end_cancel):
+                    print(f'found too old workflow {logging_detail}')
+                    job_status = "too_old"
+
+                elif username in robot_committers:
+                    continue
 
-                    yield {"job_status": "age_warning", "name": current_workflow['name'], "id": current_workflow['id'], "username": username}
+                elif current_workflow['status'] == 'on_hold':
+                    print(f'midlife warning for workflow {logging_detail}')
+                    job_status = "age_warning"
 
-                if (created_at < (now - job_life_clock)):
-                    username = get_workflow_started_by(
-                        current_workflow, headers)
-                    yield {"job_status": "too_old", "name": current_workflow['name'], "id": current_workflow['id'], "username": username}
+            if job_status:
+                yield {
+                    "job_status": "age_warning",
```

Comment:
Should this be using the `job_status` variable instead? I'm not immediately understanding how jobs get canceled otherwise.

Reply:
good catch 🤦